docs/man/man/reader.3
.TH "Reader" 3 "Sun Oct 4 2015" "Crystal Builder v 3.7.0" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Reader \- 
.SH SYNOPSIS
.br
.PP
.SH "Detailed Description"
.PP 

.SH "\fBPurpose:\fP "
.PP
.PP
.PP
.nf
/********************************************************************************\
/  Class for reading crystal parameters from a file                              \
/                                                                                \
/********************************************************************************\
.fi
.PP
.PP
\fBAuthor:\fP
.RS 4
Joseph M\&. Gonzalez
.RE
.PP
\fBVersion:\fP
.RS 4
0\&.1
.RE
.PP
\fBDate:\fP
.RS 4
Sep 27, 2015 23:16:20
.RE
.PP
\fBContact\fP 
.br
 jmgonza6@mail.usf.edu 
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBReader\fP ()"
.br
.ti -1c
.RI "\fB~Reader\fP ()"
.br
.ti -1c
.RI "void \fBfile\fP (char *pf)"
.br
.ti -1c
.RI "int \fBinsane\fP ()"
.br
.in -1c
.SS "Data Fields"

.in +1c
.ti -1c
.RI "double \fBalat\fP"
.br
.RI "\fIlatttice parameter a (Å) \fP"
.ti -1c
.RI "double \fBblat\fP"
.br
.RI "\fIlattice parameter b (Å) \fP"
.ti -1c
.RI "double \fBclat\fP"
.br
.RI "\fIlattice parameter c (Å) \fP"
.ti -1c
.RI "double \fBalpha\fP"
.br
.RI "\fIangle beteween c & b \fP"
.ti -1c
.RI "double \fBbeta\fP"
.br
.RI "\fIangle between c & a \fP"
.ti -1c
.RI "double \fBgamma\fP"
.br
.RI "\fIangle between a & b \fP"
.ti -1c
.RI "double \fBcta\fP"
.br
.RI "\fIc/a ratio, for use with Hexagonal(), h-BN(), Graphene() \fP"
.ti -1c
.RI "int \fBlbasis\fP"
.br
.RI "\fInumber of basis atoms for graphene \fP"
.ti -1c
.RI "int \fBnx\fP"
.br
.RI "\fIsupercell dimension along x \fP"
.ti -1c
.RI "int \fBny\fP"
.br
.RI "\fIsupercell dimension along y \fP"
.ti -1c
.RI "int \fBnz\fP"
.br
.RI "\fIsupercell dimension along z \fP"
.ti -1c
.RI "int \fBfractional\fP"
.br
.RI "\fIfractional coordinates, 1 = true, 0 = false \fP"
.ti -1c
.RI "int \fBgraphene\fP"
.br
.RI "\fIgraphene structure was built, 1 = true, 0 = false \fP"
.ti -1c
.RI "int \fBcustom\fP"
.br
.RI "\fIa custom structure was built, 1 = true, 0 = false \fP"
.ti -1c
.RI "int \fBresets\fP"
.br
.RI "\fIreset count \fP"
.ti -1c
.RI "int \fBrun_count\fP"
.br
.RI "\fIrun count, for debugging \fP"
.ti -1c
.RI "int \fBquit\fP"
.br
.RI "\fIquit signal \fP"
.ti -1c
.RI "int \fBview\fP"
.br
.RI "\fIrender+build signal \fP"
.ti -1c
.RI "char * \fBpfile\fP"
.br
.RI "\fIparameter file to define the crystal \fP"
.ti -1c
.RI "char * \fBout_name\fP"
.br
.RI "\fIfile name to write \fP"
.ti -1c
.RI "char * \fBfile_format\fP"
.br
.RI "\fItype of file to write \fP"
.ti -1c
.RI "char * \fBbravais\fP"
.br
.RI "\fIwhich lattice ? \fP"
.ti -1c
.RI "char * \fBlstyle\fP"
.br
.RI "\fIbravais lattice style, BCC \fP"
.ti -1c
.RI "char * \fBpd_struct\fP"
.br
.RI "\fIdefines a library structure to use \fP"
.ti -1c
.RI "char * \fBelement\fP"
.br
.RI "\fIwhich species ? \fP"
.ti -1c
.RI "std::vector< std::string > \fBelemList\fP"
.br
.RI "\fIdefine a custom stoichiometry (C,H,N,O) \fP"
.ti -1c
.RI "std::vector< int > \fBelemCount\fP"
.br
.RI "\fIdefine a custom stoichiometry (2,4,1,3) \fP"
.ti -1c
.RI "std::vector< std::vector< double > > \fBcustomBasis\fP"
.br
.RI "\fIcontainer for defining a custom basis set \fP"
.ti -1c
.RI "std::string \fBcustomName\fP"
.br
.RI "\fIchemical formula of a custom structure \fP"
.ti -1c
.RI "\fBParser\fP * \fBparser\fP"
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Reader::Reader ()"

.SS "Reader::~Reader ()"

.SH "Member Function Documentation"
.PP 
.SS "void Reader::file (char * pf)"

.SS "int Reader::insane ()"

.SH "Field Documentation"
.PP 
.SS "double Reader::alat"

.PP
latttice parameter a (Å) 
.SS "double Reader::alpha"

.PP
angle beteween c & b 
.SS "double Reader::beta"

.PP
angle between c & a 
.SS "double Reader::blat"

.PP
lattice parameter b (Å) 
.SS "char* Reader::bravais"

.PP
which lattice ? 
.SS "double Reader::clat"

.PP
lattice parameter c (Å) 
.SS "double Reader::cta"

.PP
c/a ratio, for use with Hexagonal(), h-BN(), Graphene() 
.SS "int Reader::custom"

.PP
a custom structure was built, 1 = true, 0 = false 
.SS "std::vector<std::vector<double> > Reader::customBasis"

.PP
container for defining a custom basis set 
.SS "std::string Reader::customName"

.PP
chemical formula of a custom structure 
.SS "std::vector<int> Reader::elemCount"

.PP
define a custom stoichiometry (2,4,1,3) 
.SS "char* Reader::element"

.PP
which species ? 
.SS "std::vector<std::string> Reader::elemList"

.PP
define a custom stoichiometry (C,H,N,O) 
.SS "char* Reader::file_format"

.PP
type of file to write 
.SS "int Reader::fractional"

.PP
fractional coordinates, 1 = true, 0 = false 
.SS "double Reader::gamma"

.PP
angle between a & b 
.SS "int Reader::graphene"

.PP
graphene structure was built, 1 = true, 0 = false 
.SS "int Reader::lbasis"

.PP
number of basis atoms for graphene 
.SS "char* Reader::lstyle"

.PP
bravais lattice style, BCC 
.SS "int Reader::nx"

.PP
supercell dimension along x 
.SS "int Reader::ny"

.PP
supercell dimension along y 
.SS "int Reader::nz"

.PP
supercell dimension along z 
.SS "char* Reader::out_name"

.PP
file name to write 
.SS "\fBParser\fP* Reader::parser"

.SS "char* Reader::pd_struct"

.PP
defines a library structure to use 
.SS "char* Reader::pfile"

.PP
parameter file to define the crystal 
.SS "int Reader::quit"

.PP
quit signal 
.SS "int Reader::resets"

.PP
reset count 
.SS "int Reader::run_count"

.PP
run count, for debugging 
.SS "int Reader::view"

.PP
render+build signal 

.SH "Author"
.PP 
Generated automatically by Doxygen for Crystal Builder v 3\&.7\&.0 from the source code\&.


src/util/reader.cpp
#include "reader.h"

Reader::Reader()
{
    parser = new Parser();

    out_name = NULL;
    file_format = NULL;
    bravais = NULL;

    alat=blat=clat=1;
    alpha=beta=gamma=90;

    nx=ny=nz=1;
    fractional=0;
    graphene=0;

    custom=0;
    view=0;

}

Reader::~Reader()
{
    delete parser;
}



void 
Reader::file(char *pf)
{
	std::ifstream pfile(pf);
    if (!pfile.is_open()) {printf ("Could not open the parameters file [ %s ]\n",pf);}

    double x,y,z;
    int bflag=0;


	std::istringstream iss;
	std::string keyword, line;
	std::string arg1, arg2, arg3;

    std::string blattice;
    std::string ffmt, fname;

    std::vector<std::string> keys, vals;
    std::vector<std::string> g_elemList;
    std::vector<std::string> g_elemCount;
    std::vector<std::string> basisList;


	if (DMSG)
    	fprintf (FDEBUG,"Reading from file ................: %s\n",pf);


    // search first for the required parameters
    while (std::getline(pfile,line)) {
        iss.clear();
        iss.str(line);
        iss >> keyword >> arg1 >> arg2;;

        if (keyword == "bravais") {
            bravais = parser->str2char(arg1);
        	blattice = arg1;
        } else if (keyword == "save") {
        	ffmt = arg1;
            file_format = parser->str2char(arg1);
        	if (ffmt == "vasp") {
        		fname = "POSCAR";
                out_name = parser->str2char("POSCAR");
        	} else {
                out_name = parser->str2char(arg2);
        		fname = arg2;
        	}
        } else if (keyword == "bravais_style") {
            lstyle = parser->str2char(arg1);
        }           
    }



    pfile.clear();                 // clear fail and eof bits
	pfile.seekg(0, std::ios::beg); // back to the start!

	// search for general modifications
	while (std::getline(pfile,line)) {
        iss.clear();
        iss.str(line);
        iss >> keyword >> arg1 >> arg2;

        if (keyword == "stoich") {
      		parser->get_substrings(arg1, ',', ':', g_elemList, g_elemCount);
            elemList = g_elemList;
            for (int i=0;i<g_elemCount.size();i++) {
                elemCount.push_back(parser->str2num<int>(g_elemCount[i]));
            }
			keyword = "";
        } else if (keyword == "lat_parameters") {
        	parser->get_substrings(arg1, ',', '=', keys, vals);
        	for(int i=0; i<keys.size();i++) {
        		if (keys[i]=="a") alat = parser->str2num<double>(vals[i]);
        		else if (keys[i]=="b") blat = parser->str2num<double>(vals[i]);
        		else if (keys[i]=="c") clat = parser->str2num<double>(vals[i]);
        	}
        	keyword = "";
        } else if (keyword == "lat_angles") {
        	parser->get_substrings(arg1, ',', '=', keys, vals);
        	for(int i=0; i<keys.size();i++) {
        		if (keys[i]=="alpha") alpha = parser->str2num<double>(vals[i]);
        		else if (keys[i]=="beta") beta = parser->str2num<double>(vals[i]);
        		else if (keys[i]=="gamma") gamma = parser->str2num<double>(vals[i]);
        	}
        	keyword = "";
        } else if (keyword == "super_cell") {
        	parser->get_substrings(arg1, ',', '=', keys, vals);
        	for(int i=0; i<keys.size();i++) {
        		if (keys[i]=="nx") nx = parser->str2num<int>(vals[i]);
        		else if (keys[i]=="ny") ny = parser->str2num<int>(vals[i]);
        		else if (keys[i]=="nz") nz = parser->str2num<int>(vals[i]);
        	}
        	keyword = "";
        } else if (keyword == "data_modify") {
        	if (arg1 == "fractional") {
        		fractional = 1;
        	}
        	keyword = "";
        } else if (keyword == "run") {
        	if (arg1 == "build") {
        		int build = 1;
        	} else if (arg1 == "render") {
        		view = 1;
        	} else {
        		printf ("ERROR: unknown run style [ %s ]\n",arg1.c_str());
        	}
        	keyword = "";
        } else if (keyword == "bravais_style") {
            lstyle = parser->str2char(arg1);
        }
    }

	pfile.clear();                 // clear fail and eof bits
	pfile.seekg(0, std::ios::beg); // back to the start!

   // search for library relevant parameters aside from the general modifiers
	if (blattice == "library") {
		while (std::getline(pfile,line)) {
	        iss.clear();
	        iss.str(line);
	        iss >> keyword >> arg1;

	        if (keyword == "use_prebuilt") {
	        	if (!parser->str2num<int>(arg1)) {
                    if (arg1 == "graphene") {
                        pd_struct = parser->str2char("graphene");
                        graphene=1;
                    } else if (arg1 == "hbn") {
                        pd_struct = parser->str2char("h-BN");
                    } else if (arg1 == "mdc-2h") {
                        pd_struct = parser->str2char("MDC MX2-2H");
                        if (elemList.empty()) {
                            elemList.push_back("Sn");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.684;

                        if (clat==1) clat = 5.89;
                    } else if (arg1 == "mdc-1t") {
                        pd_struct = parser->str2char("MDC MX2-1T");
                        if (elemList.empty()) {
                            elemList.push_back("Sn");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.684;

                        if (clat==1) clat = 5.89;
                    } else if (arg1 == "tmdc-2h") {
                        pd_struct = parser->str2char("TMDC MX2-2H");
                        if (elemList.empty()) {
                            elemList.push_back("Mo");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.19;

                        if (clat==1) clat = 12.3;
                    } else if (arg1 == "tmdc-1t") {
                        pd_struct = parser->str2char("TMDC MX2-1T");
                        if (elemList.empty()) {
                            elemList.push_back("Mo");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.19;

                        if (clat==1) clat = 12.3;
                    } else if (arg1 == "petn") {
                        pd_struct = parser->str2char("PETN-I");
                        elemList.clear();
                        elemCount.clear();
                        elemList.push_back("C");
                        elemList.push_back("H");
                        elemList.push_back("N");
                        elemList.push_back("O");
                        elemCount.push_back(10);
                        elemCount.push_back(16);
                        elemCount.push_back(8);
                        elemCount.push_back(24);
                    } else if (arg1 == "tatb") {
                        pd_struct = parser->str2char("TATB");
                        elemList.clear();
                        elemCount.clear();
                        elemList.push_back("C");
                        elemList.push_back("H");
                        elemList.push_back("N");
                        elemList.push_back("O");
                        elemCount.push_back(12);
                        elemCount.push_back(12);
                        elemCount.push_back(12);
                        elemCount.push_back(12);
                    } else if (arg1 == "hmx") pd_struct = parser->str2char("B-HMX");
                    
	        	} else {
                    printf ("prebuilt index = %i\n",parser->str2num<int>(arg1));
                    int index = parser->str2num<int>(arg1);
                    if (index == 1) {
                        pd_struct = parser->str2char("graphene");
                        graphene=1;
                    } else if (index == 2) {
                        pd_struct = parser->str2char("h-BN");
                    } else if (index == 3) {
                        pd_struct = parser->str2char("MDC MX2-2H");
                        if (elemList.empty()) {
                            elemList.push_back("Sn");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.684;

                        if (clat==1) clat = 5.89;
                    } else if (index == 4) {
                        pd_struct = parser->str2char("MDC MX2-1T");
                        if (elemList.empty()) {
                            elemList.push_back("Sn");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.684;

                        if (clat==1) clat = 5.89;
                    } else if (index == 5) {
                        pd_struct = parser->str2char("TMDC MX2-2H");
                        if (elemList.empty()) {
                            elemList.push_back("Mo");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.19;

                        if (clat==1) clat = 12.3;
                    } else if (index == 6) {
                        pd_struct = parser->str2char("TMDC MX2-1T");
                        if (elemList.empty()) {
                            elemList.push_back("Mo");
                            elemList.push_back("S");
                            elemCount.push_back(1);
                            elemCount.push_back(2);
                        } 
                        if (alat==1) alat = 3.19;

                        if (clat==1) clat = 12.3;
                    } else if (index == 7) {
                        pd_struct = parser->str2char("PETN-I");
                        elemList.clear();
                        elemCount.clear();
                        elemList.push_back("C");
                        elemList.push_back("H");
                        elemList.push_back("N");
                        elemList.push_back("O");
                        elemCount.push_back(10);
                        elemCount.push_back(16);
                        elemCount.push_back(8);
                        elemCount.push_back(24);
                    } else if (index == 8) {
                        pd_struct = parser->str2char("TATB");
                        elemList.clear();
                        elemCount.clear();
                        elemList.push_back("C");
                        elemList.push_back("H");
                        elemList.push_back("N");
                        elemList.push_back("O");
                        elemCount.push_back(12);
                        elemCount.push_back(12);
                        elemCount.push_back(12);
                        elemCount.push_back(12);
                    } else if (index == 9) {
                        pd_struct = parser->str2char("B-HMX");
                    }
                }
	        		
	        } else if (keyword == "graphene_basis") {
                lbasis = parser->str2num<int>(arg1);
            }
	        keyword = "";
	    }
	} else if (blattice == "cubic") {
		blat = clat = alat;
		alpha = beta = gamma = 90;
	} else if (blattice == "tetragonal") {
		blat = alat;
		alpha = beta = gamma = 90;
	} else if (blattice == "ortho") {
		alpha = beta = gamma = 90;
	} 


    // if a custom crystal was requested, then search for the basis atoms
    if (blattice == "custom") {
        custom = 1;
        pfile.clear();                 // clear fail and eof bits
        pfile.seekg(0, std::ios::beg); // back to the start!

        std::vector<double> row(3);

        bflag = 0;
        while (std::getline(pfile,line)) {
            iss.clear();
            iss.str(line);
            iss >> arg1 >> arg2 >> arg3;
            if (arg1 == "begin_basis") {
                arg1 = "";
                while (std::getline(pfile,line)) {
                    iss.clear();
                    iss.str(line);
                    iss >> arg1 >> arg2 >> arg3;
                    // stop reading if we have all the atoms
                    if (arg1 == "end_basis") {bflag=1;break;}

                    row[0] = parser->str2num<double>(arg1);
                    row[1] = parser->str2num<double>(arg2);
                    row[2] = parser->str2num<double>(arg3);
                    
                    arg1 = "";
                    customBasis.push_back(row);
                }
                if (bflag) break;
            }
        }

        // create a name for the structure based on its stoichiometry
        std::string unit;
        for (int i=0;i<g_elemList.size();i++) {
            // if there is only 1 of a species, print only the element symbol
            if (parser->str2num<int>(g_elemCount[i])==1) unit = g_elemList[i];
            else unit = g_elemList[i] + g_elemCount[i];
            customName += unit;
        }
    }

    if (DMSG) {
        printf ("blattice = %s\n",bravais);
        printf ("ifmt = %s\n",file_format);
        printf ("fname = %s\n",out_name);
        printf ("latticeStyle = %s\n",lstyle);
        for (int i=0;i<elemList.size();i++)
            printf ("stoich = %s %i\n",elemList[i].c_str(),elemCount[i]);


        printf ("nx    = %i\n",nx);
        printf ("ny    = %i\n",ny);
        printf ("nz    = %i\n",nz);
        printf ("alat  = %f\n",alat);
        printf ("blat  = %f\n",blat);
        printf ("clat  = %f\n",clat);
        printf ("alpha = %f\n",alpha);
        printf ("beta  = %f\n",beta);
        printf ("gamma = %f\n",gamma);

    }
}

int 
Reader::insane()
{

    // check for different parameters if the structure was custom
    if (custom) {

        if (!alat) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> lattice parameter 'a' not defined!\n",__LINE__);
            return 1;
        }

        if (!blat) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> lattice parameter 'b' not defined!\n",__LINE__);
            return 1;
        }

        if (!clat) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> lattice parameter 'c' not defined!\n",__LINE__);
            return 1;
        }

        if (!alpha) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> lattice angle 'alpha' not defined!\n",__LINE__);
            return 1;
        }
    
        if (!beta) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> lattice angle 'beta' not defined!\n",__LINE__);
            return 1;
        }
        if (!gamma) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> lattice angle 'gamma' not defined!\n",__LINE__);
            return 1;
        }
        
        if (elemList.empty()) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> no atomic species were defined!\n",__LINE__);
            return 1;
        }
        if (customBasis.empty()) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> no basis atoms were defined!\n",__LINE__);
            return 1;
        }

        if (out_name == NULL && strcmp(file_format,"vasp (POSCAR)")!=0) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> no file name given for non VASP format!\n",__LINE__);
            return 1;
        }
        return 0;

    } else {
        // deal with the prebuilt structures first
        if (graphene) {
            if (!alat) {
                fprintf (EMESG,"Reader::insane()::ERROR\n");
                fprintf (EMESG,"  line %d -> lattice parameter 'a' was not defined!\n",__LINE__);
                return 1;
            }
            if (strcmp(file_format,"vasp")==0 
                || strcmp(file_format,"lammps")==0 
                || strcmp(file_format,"dmol")==0) { return 0; }
        }

        if (strcmp(bravais,"library")==0 && strcmp(file_format,"vasp")==0) {return 0;}
        if (strcmp(bravais,"library")==0 && strcmp(file_format,"lammps")==0) {return 0;}
        if (strcmp(bravais,"library")==0 && strcmp(file_format,"dmol ")==0) {return 0;}

        if (strcmp(file_format,"vasp")==0 && alat !=0 && !elemList.empty() && bravais != NULL) {return 0;}

        // check that ALL necessary gui parameters are set
        // if not, print which one was not set and return 1
        if (file_format==NULL) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> file format wasnt defined!\n",__LINE__);
            return 1;
        }
        // check for lattice specific values
        if (!bravais) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> bravais lattice was not defined!\n",__LINE__);
            return 1;
        } else {
            if (strcmp(bravais,"cubic")==0 || strcmp(bravais,"hexagonal")==0) {
                if (!alat) {
                    fprintf (EMESG,"Reader::insane()::ERROR\n");
                    fprintf (EMESG,"  line %d -> lattice parameter 'a' was not defined!\n",__LINE__);
                    return 1;
                }
            }

            if (strcmp(bravais,"tetragonal")==0) {
                if (!alat) {
                    fprintf (EMESG,"Reader::insane()::ERROR\n");
                    fprintf (EMESG,"  line %d -> lattice parameter 'a' was not defined!\n",__LINE__);
                    return 1;
                }
                if (!clat) {
                    fprintf (EMESG,"Reader::insane()::ERROR\n");
                    fprintf (EMESG,"  line %d -> lattice parameter 'c' was not defined!\n",__LINE__);
                    return 1;
                }
            }

            if (strcmp(bravais,"ortho")==0) {
                if (!alat) {
                    fprintf (EMESG,"Reader::insane()::ERROR\n");
                    fprintf (EMESG,"  line %d -> lattice parameter 'a' was not defined!\n",__LINE__);
                    return 1;
                }
                if (!blat) {
                    fprintf (EMESG,"Reader::insane()::ERROR\n");
                    fprintf (EMESG,"  line %d -> lattice parameter 'b' was not defined!\n",__LINE__);
                    return 1;
                }
                if (!clat) {
                    fprintf (EMESG,"Reader::insane()::ERROR\n");
                    fprintf (EMESG,"  line %d -> lattice parameter 'c' was not defined!\n",__LINE__);
                    return 1;
                }
            }
        }
        if (out_name == NULL && strcmp(file_format,"vasp")!=0) {
            fprintf (EMESG,"Reader::insane()::ERROR\n");
            fprintf (EMESG,"  line %d -> no file name given for non VASP format!\n",__LINE__);
            return 1;
        }
        return 0;
    }
}
